src/duqtools/ids/_mapping.py

```diff
@@ -22,69 +24,82 @@ def insert_re_caret_dollar(string: str) -> str:
-    """Recursive defaultdict."""
-    return defaultdict(recursive_defaultdict)
-
-
-def defaultdict_to_dict(ddict: defaultdict):
-    """defaultdict_to_dict, turns a nested defaultdict into a nested dict.
-
-    Parameters
-    ----------
-    ddict : defaultdict
-        ddict
-    """
-
-    # Convert members to dict first
-    for k, v in ddict.items():
-        if isinstance(v, defaultdict):
-            ddict[k] = defaultdict_to_dict(v)
-    # Finally convert self to dict
-    return dict(ddict)
-
-
 class IDSMapping(Mapping):
 
     def __init__(self, ids, exclude_empty: bool = True):
         self._ids = ids
         self.exclude_empty = exclude_empty
 
-        # All fields in the core profile in a single dict
-        self.flat_fields: dict = {}
-
-        # All fields, in the core profile in a nested dict
-        self.fields: dict = defaultdict(recursive_defaultdict)
+        # All available data fields are stored in this set.
+        self._keys: Set[str] = set()
 
         self.dive(ids, [])
-        self.fields = defaultdict_to_dict(self.fields)
 
     def __repr__(self):
         s = f'{self.__class__.__name__}(\n'
-        for key in self.fields.keys():
+        for key in self._keys:
             s += f'  {key} = ...\n'
         s += ')\n'
 
         return s
 
     def __getitem__(self, key: str):
-        try:
-            return self.flat_fields[key]
-        except KeyError:
-            pass
-        try:
-            return self.fields[key]
-        except KeyError:
-            raise
+        if key not in self._keys:
+            raise KeyError(key)
+
+        *parts, attr = key.split('/')
+
+        pointer = self._ids
+
+        for part in parts:
+            try:
+                i = int(part)
+            except ValueError:
+                pointer = getattr(pointer, part)
+            else:
+                pointer = pointer[i]
+
+        return getattr(pointer, attr)
+
+    def __setitem__(self, key: str, value: np.ndarray):
+        if key not in self._keys:
+            return f'Cannot set non-existant key: {key}'
+
+        *parts, attr = key.split('/')
+
+        pointer = self._ids
+
+        for part in parts:
+            try:
+                i = int(part)
+            except ValueError:
+                pointer = getattr(pointer, part)
+            else:
+                pointer = pointer[i]
+
+        setattr(pointer, attr, value)
 
     def __iter__(self):
-        yield from self.fields
+        yield from self._keys
 
     def __len__(self):
-        return len(self.fields)
+        return len(self._keys)
+
+    def sync(self, handle: ImasHandle):
+        """Synchronize updated data back to IMAS db entry.
+
+        Shortcut for 'put' command.
+
+        Parameters
+        ----------
+        handle : ImasHandle
+            Points to an IMAS db entry of where the data should be written.
+        """
+        with handle.open() as f:
+            self._ids.put(db_entry=f)
```

Comment:
I guess we have to write the whole thing back, not just the updated values? makes sense for the legacy IMAS structure, but perhaps for the hdf5 this can eventually be improved (just updating the updated fields).

Reply:
Agree, we do the same thing when creating the data. I don't know any better way at the moment.